google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/DelayTransactionStartUntilFirstWriteMockServerTest.java

```diff
@@ -0,0 +1,575 @@
+                      "show variable %sdelay_transaction_start_until_first_write", prefix)))) {
+        assertTrue(resultSet.next());
+        assertFalse(resultSet.getBoolean(0));
+        assertFalse(
+            resultSet.getBoolean(
+                prefix.toUpperCase() + "DELAY_TRANSACTION_START_UNTIL_FIRST_WRITE"));
+        assertFalse(resultSet.next());
+      }
+    }
+    try (Connection connection = createConnection(";delayTransactionStartUntilFirstWrite=true")) {
+      assertTrue(connection.isDelayTransactionStartUntilFirstWrite());
+      connection.execute(
+          Statement.of(
+              String.format("set %sdelay_transaction_start_until_first_write=false", prefix)));
+      assertFalse(connection.isDelayTransactionStartUntilFirstWrite());
+    }
+  }
+
+  @Test
+  public void testDefaultUsesRealTransactions() {
+    try (Connection connection = createConnection()) {
+      executeRandomQuery(connection);
+      connection.commit();
+
+      assertEquals(1, mockSpanner.countRequestsOfType(ExecuteSqlRequest.class));
+      ExecuteSqlRequest request = mockSpanner.getRequestsOfType(ExecuteSqlRequest.class).get(0);
+      assertTrue(request.hasTransaction());
+      assertTrue(request.getTransaction().hasBegin());
+      assertTrue(request.getTransaction().getBegin().hasReadWrite());
+      assertEquals(1, mockSpanner.countRequestsOfType(CommitRequest.class));
+    }
+  }
+
+  @Test
+  public void testSingleQuery() {
+    try (Connection connection = createConnection()) {
+      connection.setDelayTransactionStartUntilFirstWrite(true);
+
+      for (boolean commit : new boolean[] {true, false}) {
+        executeRandomQuery(connection);
+        if (commit) {
+          connection.commit();
+        } else {
+          connection.rollback();
+        }
+
+        assertEquals(1, mockSpanner.countRequestsOfType(ExecuteSqlRequest.class));
+        ExecuteSqlRequest request = mockSpanner.getRequestsOfType(ExecuteSqlRequest.class).get(0);
+        assertFalse(request.hasTransaction());
+        assertEquals(0, mockSpanner.countRequestsOfType(CommitRequest.class));
+        assertEquals(0, mockSpanner.countRequestsOfType(RollbackRequest.class));
+
+        mockSpanner.clearRequests();
+      }
+    }
+  }
+
+  @Test
+  public void testTwoQueries() {
+    try (Connection connection = createConnection()) {
+      connection.setDelayTransactionStartUntilFirstWrite(true);
+
+      for (boolean commit : new boolean[] {true, false}) {
+        executeRandomQuery(connection);
+        executeRandomQuery(connection);
+        if (commit) {
+          connection.commit();
+        } else {
+          connection.rollback();
+        }
+
+        assertEquals(2, mockSpanner.countRequestsOfType(ExecuteSqlRequest.class));
+        assertFalse(mockSpanner.getRequestsOfType(ExecuteSqlRequest.class).get(0).hasTransaction());
+        assertFalse(mockSpanner.getRequestsOfType(ExecuteSqlRequest.class).get(1).hasTransaction());
```

Comment:
Why does the second query has a transaction? I would expect both of them to be using a singleUse (because there were no writes)?

Reply:
I think you misread the assertion... Or I'm missing something. The line says `assertFalse(... hasTransaction())`.

So your assumption is right. Both are single-use read-only.